subiquity/controllers/installprogress.py

```diff
@@ -40,10 +49,168 @@ class InstallState:
     ERROR = -1
 
 
+class WaitForCurtinEventsTask(BackgroundTask):
+
+    def __init__(self, controller):
+        self.controller = controller
+        self.waited = 0.0
+
+    def start(self):
+        pass
+
+    def _bg_run(self):
+        while self.controller._event_indent and self.waited < 5.0:
+            time.sleep(0.1)
+            self.waited += 0.1
+        log.debug("waited %s seconds for events to drain", self.waited)
+
+    def end(self, observer, fut):
+        # Will raise if command failed:
+        fut.result()
+        self.controller._install_event_start("installing snaps")
+        observer.task_succeeded()
+
+
+class DownloadSnapTask(BackgroundTask):
+    # tl;dr: "snap download" makes me slightly sad
+    #
+    # When the user requests that a snap be installed, we need to
+    # download the snap and associated assertion, put those in
+    # /var/lib/snapd/seed/{snaps,assertions}/ and add it
+    # /var/lib/snapd/seed/seed.yaml (all in the target system).  The
+    # data in the seed.yaml is the name of the snap, the name of the
+    # downloaded file and the channel to track.
+    #
+    # "snap download" downloads the snap and assertion for us but the
+    # part that makes me slightly sad is that it doesn't tell us the
+    # name of the files it downloaded in any reasonable way. We could
+    # make some assumptions -- they're
+    # ${SNAP_NAME}_${REVISION}.{snap,assert} and I don't think
+    # SNAP_NAME or REVISION can contain underscores -- but this feels
+    # a bit icky. So what I do is run each snap download in a separate
+    # directory -- /var/lib/snapd/seed/tmp/$SNAP_NAME/ -- so
+    # UpdateSnapSeed can be sure that the only snap/assertion in
+    # /var/lib/snapd/seed/tmp/$SNAP_NAME/ is indeed the snap/assertion
+    # for $SNAP_NAME.
+    #
+    # So after all the DownloadSnapTasks have run, the directory tree
+    # looks like this:
+    #
+    # /var/lib/snapd/seed/
+    #         seed.yaml -- seed that came from the base system
+    #         snaps/
+    #           [snaps from base seed]
+    #         assertions/
+    #           [assertions from base seed]
+    #         tmp/
+    #           $SNAP1/
+    #             $SNAP1_$REV1.snap
+    #             $SNAP1_$REV1.assert
+    #           $SNAP2/
+    #             $SNAP2_$REV2.snap
+    #             $SNAP2_$REV2.assert
```

Comment:
Would it be better to have a tempdir per task?  Rather than share /var/lib/snapd/seed/tmp and determine when it is OK to remove the dir, each "download" task would create one, and once it's relocated the snap/assert files, then clean itself up.?

One thing to note is that /tmp is not safe due to systemd tmpfile removal, we have a wrapper/context manager for using /run/<project name>/tmp  which would be safe from systemd.

https://git.launchpad.net/cloud-init/tree/cloudinit/temp_utils.py#n83

Reply:
We can't store the snaps in /tmp or /run/tmp even temporarily because that risks running the live session out of memory (snaps can be big!).

There are clearly other ways we could skin this cat, we could update the seed after each snap download, we could store the snap file name in some data structure rather than in the filesystem or some other things I haven't thought of yet, but I think this is OK? If snap download starts downloading something other than the snap and the assertion, we're definitely going to have to change this code.